Automate t1-t6 of mom dynamic resource test

dinesh1306 opened the pull request:
<!--- Please review your changes in preview mode -->
<!--- Provide a general summary of your changes in the Title above -->

#### Bug/feature Description
* Automate test cases t1-t6 from mom dynamic resource test suite.

#### Affected Platform(s)
* All

#### Solution Description
* Added test for mom dynamic resource. 
* t1.host level consumable size resource with mom dynamic value string
* t2.host level float resource with mom dynamic value string
* t3.host level string resource with mom dynamic value integer
* t4.host level string_array resource with mom dynamic value float
* t5.host level string_array resource with mom dynamic value string
* t6.host level string_array resource with mom dynamic value “This is a test”


#### Checklist:
<!--- Use the preview button to see the checkboxes/links properly. -->
<!--- Go over the following points, and put an `x` (without spaces around it) in the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have joined the **[pbspro community forum](http://community.pbspro.org/)**.
- [x] My pull request contains a **single, signed** commit. See **[setting up gpg signature](https://pbspro.atlassian.net/wiki/display/DG/Signing+Your+Git+Commits)**.
- [x] My code follows the **[coding style](https://pbspro.atlassian.net/wiki/display/DG/Coding+Standards)** of this project.
- [ ] My change requires project documentation. See **[required documentation checklist](https://pbspro.atlassian.net/wiki/display/DG/Checklist+for+Developing+Features+and+Bug+Fixes)** for details.
   - [ ] I have added documentation in the **[project documentation area](https://pbspro.atlassian.net/wiki/display/PD)**.
- [ ] I have added new **PTL test(s) to my commit**. (See **[using PTL for testing](https://pbspro.atlassian.net/wiki/display/DG/Using+PTL+for+Testing)**) *(or)*
   - [ ] I have added  **manual test(s) to this pull request and explained why PTL is not appropriate** for this case.
- [x] All new and existing automated tests have passed. (See **[running automated PTL tests](https://pbspro.atlassian.net/wiki/display/DG/PTL+Quick+Start+Guide)**).
- [x] I have attached **test logs to this pull request** as evidence of testing/verification.
[result.txt](https://github.com/PBSPro/pbspro/files/1829680/result.txt)



__***For further information please visit the [Developer Guide Home](https://pbspro.atlassian.net/wiki/display/DG/Developer+Guide+Home).***__

anamikau commented on test/tests/functional/pbs_mom_dynamic_resource_ext.py:
```diff
@@ -0,0 +1,212 @@
+
+        filename = "foobar"
+        tmpfilepath = os.path.join(os.sep, "tmp", filename)
+        with open(tmpfilepath, "w") as fd:
```
use mkstemp to create a file or dshutils open function. for system like ICE or CRAY mom is not on the same host.

anamikau commented on test/tests/functional/pbs_mom_dynamic_resource_ext.py:
```diff
@@ -0,0 +1,212 @@
+        # The job should run
+        self.server.expect(JOB, {'job_state': 'R'}, id=jid)
+
+        # Submit a job that requests mom dynamic resource
```
update the comment to reflect job is asking a different resource

dinesh1306 commented:
@anamikau , i attached result with this PR (result.txt). now i am attaching result after addressed your comment.
[result.log](https://github.com/PBSPro/pbspro/files/1837329/result.log)

dinesh1306 commented:
@anamikau , 
I have addressed your comments , please take a look.

dinesh1306 commented:
@anamikau , i update the summary of PR.
Please take a look.

dinesh1306 commented:
@PBSPro/pbspro-maintainers Please review the PR.

bhroam commented on test/tests/functional/pbs_mom_dynamic_resource_ext.py:
```diff
@@ -0,0 +1,212 @@
+                            resc_type, resc_flag,
+                            script_body):
+        """
+        helper function for create mom_dyn_res.
```
I'd remove the underscores.  The real PBS keyword is mom_resources

bhroam commented on test/tests/functional/pbs_mom_dynamic_resource_ext.py:
```diff
@@ -0,0 +1,212 @@
+        self.server.expect(JOB, {'job_state': 'Q'}, id=jid)
+
+        # Check for the expected log message in scheduler logs
+        c = "Insufficient amount of resource: foo (1kb != (null))"
```
Any time you see '(null)' in a string from PBS, then it is a bug.  It means a NULL pointer was passed into printf().  This is not something you should be testing for in a PTL test.

bhroam commented on test/tests/functional/pbs_mom_dynamic_resource_ext.py:
```diff
@@ -0,0 +1,212 @@
+        self.server.expect(JOB, {'job_state': 'Q'}, id=jid)
+
+        # Check for the expected log message in scheduler logs
+        c = "Insufficient amount of resource: foo (2 != (null))"
```
Same as above, this is a bug in PBS

bhroam commented on test/tests/functional/pbs_mom_dynamic_resource_ext.py:
```diff
@@ -0,0 +1,212 @@
+
+        # Check for the expected log message in scheduler logs
+        c = "Insufficient amount of resource: foo (abc != 3)"
+        self.scheduler.log_match(c)
```
Checking the log message is less efficient than the comment message.  Do an expect() on the job comment instead.

bhroam commented on test/tests/functional/pbs_mom_dynamic_resource_ext.py:
```diff
@@ -0,0 +1,212 @@
+    def test_t4(self):
+        """
+        Test for host level string_array resource
+        with float mom dynamic value
```
I don't think you are not really testing what you think you are testing.  What you are testing is that the scheduler is correctly not matching the string 'red' and the string '3.14'.

bhroam commented on test/tests/functional/pbs_mom_dynamic_resource_ext.py:
```diff
@@ -0,0 +1,212 @@
+
+    def test_t5(self):
+        """
+        Tets for host level string_array resource
```
any particular reason you are using a string_array resource?  You're really treating it as a string.

bhroam commented:
I don't know if you should spend time writing mom_resources tests.  The functionality has been replaced with a mom periodic hooks.  The option still exists, but could be removed at any time.

anamikau commented:
Hey @bhroam, you are right but I haven't heard of dropping that feature in 18.x at this point. These tests would still add value in testing 18.1 as there are customers still using dynamic resources. A critical bug was found earlier in server dynamic resource.

bhroam commented:
I'm incorrect.  I did a search for mom_resources in my PDF viewer and nothing came up.  It appears my PDF viewer is lame.  I tried a different viewer and mom_resources is still in the guides.  Maybe I'm wrong about it being deprecated yet.  It sure should be.  In any case, you are right, these tests should be added.  As a note, server_dyn_res and mom_resources are very different things.  server_dyn_res is heavily used.  mom_resources most likely isn't because of the scalability issues it entails.

dinesh1306 commented:
@bhroam , i have incorporate your comments. Please take a look

bhroam commented on test/tests/functional/pbs_mom_dynamic_resource_ext.py:
```diff
@@ -60,98 +61,27 @@ def create_mom_dyn_resc(self, resc_name,
-        self.server.expect(JOB, {'job_state': 'Q'}, id=jid)
-
-        # Check for the expected log message in scheduler logs
-        c = "Insufficient amount of resource: foo (1kb != (null))"
```
Did you file this bug?  It is a product bug.

bhroam replied:
Thanks for filing the bug!

anamikau replied:
@bhroam and @dinesh1306 should not we keep the test but comment the log message due to PP-1263? And these tests as reference to PP-1263.

bhroam replied:
The tests have value as negative test cases.  The problem with including them now is that depending how PP-1263 is fixed, the tests will have to change.  Probably the best way the fix PP-1263 is to reject the resource at the time we query the mom.  The test as written will not work for that.  We might want to wait until PP-1263 is fixed before adding these tests.

anamikau replied:
@bhroam you are right that functionality might change which will change the test but that should not be a reason not to include this test for 18.x. We have found an issue where negative values has caused scheduler to behave differently. At least having this test around for 18.x will help catching such issues (Today it is not crashing it might crash tomorrow with the same test). (PBS behavior might change but objective of the test would remain the same).

bhroam replied:
The test as it stands tests for the buggy behavior.  I do not think we should include that.  If we try and fix the test, we'd need to know what the non-buggy behavior is.  We won't know that until the bug is fixed.

anamikau replied:
okay thanks

dinesh1306 commented on test/tests/functional/pbs_mom_dynamic_resource_ext.py:
```diff
@@ -181,32 +111,38 @@ def test_t5(self):
+        c = "Can Never Run: Insufficient amount of resource:"
+        c += " foo (white != red,green,blue)"
+        self.server.expect(JOB, {'job_state': 'Q', 'comment': c},
+                           id=jid, attrop=PTL_AND)
```
Yes , I have checked this also. I submitted job with value 'red' and checked that job is in running state.Code is present above these lines.

bhroam replied:
Ah OK, it didn't show up because I was looking at a diff.  Thanks for pointing it out.

bhroam commented:
Looks good.  I sign off.  Once @anamikau takes another look, squash and rebase for the merge.

bhroam commented on test/tests/functional/pbs_mom_dynamic_resource_ext.py:
```diff
@@ -0,0 +1,212 @@
+        tmpfilepath = os.path.join(os.sep, "tmp", filename)
+        with open(tmpfilepath, "w") as fd:
+            fd.write(script_body)
+        self.du.chmod(path=tmpfilepath, mode=0755)
```
Now that I'm looking at this again, dshutils now provides the ability to create temp files.  Use self.du.create_temp_file().  You can even provide it the body of your script, so you don't have to do the write.  It'll do everything for you.  It returns the path to the file.

dinesh1306 replied:
@bhroam , i think you are looking into first commit. in second commit i am using "self.du.mkstemp" according to the Anamika's comment. please let me know if "self.du.mkstemp" is ok or i need to change to "self.du.create_temp_file()".

bhroam replied:
Oh, I must have been looking at the older commit.  I think self.du.create_temp_file() is a better solution because it will do everything you need.  You won't need to open the file and write to the file.  You just pass create_temp_file() the body of the file you want, and it returns the path to the newly created temp file.  It even closes the file descriptor that mkstemp() returns.

dinesh1306 commented:
@bhroam  , i have addressed your comment. please take look.

dinesh1306 commented:
rebased and squashed.Ready for merge

dinesh1306 commented:
[res_pbs_mom_dynamic_resource_ext.txt](https://github.com/PBSPro/pbspro/files/1918991/res_pbs_mom_dynamic_resource_ext.txt)

final execution details
